### PR TITLE
gh-102542 Refactor the mime audio module

### PR DESCRIPTION
**Pull request by JosephSBoyle**
Refactor the MIME audio module, the mechanism for attempting to infer a subtype is currently quite verbose.

Notes:
- `fakefile` is not used in any of the functions it is passed to, hence it has been removed.
- Slicing the header (`hrd = data[:512]`) isn't necessary so has been removed.

* Issue: gh-102542

### CODE REVIEW

File: Lib/email/mime/audio.py [hunk 1]
```diff
@@ -36,65 +35,20 @@ def __init__(self, _audiodata, _subtype=None,
         constructor, which turns them into parameters on the Content-Type
         header.
         """
-        if _subtype is None:
-            _subtype = _what(_audiodata)
-        if _subtype is None:
-            raise TypeError('Could not find audio MIME subtype')
+        _subtype = _subtype or _infer_subtype(_audiodata)
```
Comment (terryjreedy) [inline Lib/email/mime/audio.py:38]: This changes the behavior if _subtype is '' (or any other non-None null value).   Someone other than me needs to agree with this.

**Comment by terryjreedy** [comment]
@warsaw This PR mostly replaces the section you committed April '22.  I believe that replacement logic is correct.

Joseph, have you checked the coverage of the mime audio test?

File: Lib/email/mime/audio.py (hunk 1 above)
Comment (JosephSBoyle) [reply Lib/email/mime/audio.py:38]: That's correct. Falsey values other than `None` such as: `0`, `False`, `''` and empty collections will all fall back to the second `or` operand: `_infer_subtype(_audiodata)`.

**Comment by JosephSBoyle** [comment]
Force pushed to correct a mistake in the commit message.

**Comment by AlexWaygood** [comment]
> None of the tests execute this path as far as I could tell from putting a breakpoint in and running the tests.

That implies that we should perhaps first work on improving test coverage, before undertaking a risky refactoring of this module.

("Risky" only in the sense that all refactorings are risky if test coverage is low :)

**Comment by JosephSBoyle** [comment]
> That implies that we should perhaps first work on improving test coverage, before undertaking a risky refactoring of this module.

You're right as usual;) Turns out adding the tests was easier than I first thought.

Apologies for rebasing rather than merging, I'm still working on using that workflow @AlexWaygood

**Comment by JosephSBoyle** [comment]
Update: I was incorrect about this path not being tested - there are in fact tests for this, see: `TestMIMEAudio::test_guess_minor_type`.

As such I've deleted the redundant ones.

**Comment by JosephSBoyle** [comment]
@terryjreedy could you have a re-review when you get a chance please.

I think the minimum set of changes for this PR would be simply removing the unused `fakefile` that is passed to each function as this isn't used whatsoever. If the other changes don't seem worthwhile let me know and they can be reverted.

**Review by terryjreedy — COMMENTED** [review]
I have nothing to add, which was that I will not approve the change as this is outside my expertise.

**Comment by JosephSBoyle** [comment]
Hia @AlexWaygood. Who I should request a review from on this? I have the feeling it may have slipped through people's todos!

Cheers

File: Lib/email/mime/audio.py (hunk 1 above)
Comment (warsaw) [reply Lib/email/mime/audio.py:38]: I personally prefer more explicit tests than just "falseiness" when a more explicit type is indicated.  In this case `_subtype=None` defaults to None to mean that this argument wasn't given.

**Comment by warsaw** [comment]
I've provided some additional thoughts on #102542

**Comment by JosephSBoyle** [comment]
Closing as we've decided to leave the `@rule` decorator for extensibility reasons; see the issue for discussion.

**Pull request closed without merging**